File: backend/backend.API/Modules/Cars/Domain/Car.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace backend.API.Modules.Cars.Domain;

public enum PanelStatus
{
     Orijinal,
     Boyali,
     Değişmiş
}

public enum VitesTipi
{
    Düz,
    Otomatik,
    YariOtomatik
}

public enum YakitTipi
{
     Benzin,
     Dizel,
     Elektrik,
     Hibrit,
     LPG,
    Benzin_LPG,
}

public enum KasaTipi
{
    Sedan,
    Hatchback,
    SUV,
    Crossover,
    Coupe,
    Cabrio,
    StationWagon, // Kombi
    Minivan,
    Pickup,
    Van           // Panelvan
}

public enum CekisTuru
{
     ÖndenÇekiş,
    Arkadanİtiş,
    DörtÇeker,
    DörtcarpiDört
}

public enum AracDurumu
{
    Sifir,
     İkinciEl
}

public enum Kimden
{
     Sahibinden,
     Galeriden
}

public class BoyaliveDegisen
{
    [BsonElement("rightRearFender")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SağArkaÇamurluk { get; set; } = PanelStatus.Orijinal;  // Sağ Arka Çamurluk

    [BsonElement("rearHood")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus ArkaKaput { get; set; } = PanelStatus.  Orijinal;         // Arka Kaput

    [BsonElement("leftRearFender")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SolArkaÇamurluk { get; set; } = PanelStatus.  Orijinal;   // Sol Arka Çamurluk

    [BsonElement("rightRearDoor")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SağArkaKapi { get; set; } = PanelStatus.  Orijinal;    // Sağ Arka Kapı

    [BsonElement("rightFrontDoor")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SağÖnKapi { get; set; } = PanelStatus.  Orijinal;   // Sağ Ön Kapı

    [BsonElement("roof")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus Tavan { get; set; } = PanelStatus.  Orijinal;             // Tavan

    [BsonElement("leftRearDoor")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SolArkaKapi { get; set; } = PanelStatus.  Orijinal;     // Sol Arka Kapı

    [BsonElement("leftFrontDoor")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SolÖnKapi { get; set; } = PanelStatus.  Orijinal;    // Sol Ön Kapı

    [BsonElement("rightFrontFender")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SağÖnÇamurluk { get; set; } = PanelStatus.  Orijinal; // Sağ Ön Çamurluk

    [BsonElement("engineHood")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus MotorKaputu { get; set; } = PanelStatus.  Orijinal;       // Motor Kaputu

    [BsonElement("leftFrontFender")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus SolÖnÇamurluk { get; set; } = PanelStatus.  Orijinal;  // Sol Ön Çamurluk

    [BsonElement("frontBumper")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus ÖnTampon { get; set; } = PanelStatus.  Orijinal;      // Ön Tampon

    [BsonElement("rearBumper")]
    [BsonRepresentation(BsonType.String)]
    public PanelStatus  ArkaTampon { get; set; } = PanelStatus.  Orijinal;       // Arka Tampon
}

public class Car
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = ObjectId.GenerateNewId().ToString();

    [BsonElement("brand")]
    public string Marka { get; set; } = string.Empty;                         // Marka

    [BsonElement("series")]
    public string Seri { get; set; } = string.Empty;                        // Seri

    [BsonElement("model")]
    public string Model { get; set; } = string.Empty;                         // Model

    [BsonElement("year")]
    public int Yil { get; set; }                                              // Yıl

    [BsonElement("price")]
    public decimal Fiyat { get; set; }                                         // Fiyat

    [BsonElement("kilometer")]
    public int Kilometre { get; set; }                                         // Kilometre

    [BsonElement("transmissionType")]
    [BsonRepresentation(BsonType.String)]
    public VitesTipi VitesTipi { get; set; }                     // Vites Tipi

    [BsonElement("fuelType")]
    [BsonRepresentation(BsonType.String)]
    public YakitTipi YakitTipi { get; set; }                                     // Yakıt Tipi

    [BsonElement("bodyType")]
    [BsonRepresentation(BsonType.String)]
    public KasaTipi KasaTipi { get; set; }                                     // Kasa Tipi

    [BsonElement("color")]
    public string Renk { get; set; } = string.Empty;                         // Renk

    [BsonElement("engineVolume")]
    public double MotorHacmi { get; set; }                                      // Motor Hacmi (cc)

    [BsonElement("enginePower")]
    public int MotorGucu { get; set; }                                       // Motor Gücü (HP)

    [BsonElement("driveType")]
    [BsonRepresentation(BsonType.String)]
    public CekisTuru Cekis { get; set; }                                   // Çekiş

    [BsonElement("vehicleCondition")]
    [BsonRepresentation(BsonType.String)]
    public AracDurumu AracDurumu { get; set; }                     // Araç Durumu

    [BsonElement("avgFuelConsumption")]
    public double OrtalamaYakitTuketim { get; set; }                             // Ort. Yakıt Tüketimi (L/100km)

    [BsonElement("fuelTankCapacity")]
    public int YakitDeposu { get; set; }                                  // Yakıt Deposu (L)

    [BsonElement("isHeavilyDamaged")]
    public bool AgirHasarKaydi { get; set; } = false;                       // Ağır Hasarlı

    [BsonElement("isEligibleForTrade")]
    public bool TakasaUygun { get; set; } = false;                     // Takasa Uygun

    [BsonElement("sellerType")]
    [BsonRepresentation(BsonType.String)]
    public Kimden Kimden { get; set; }                                 // Kimden

    [BsonElement("images")]
    public List<string> Resimler { get; set; } = new();                         // Görseller

    [BsonElement("location")]
    public string Konum { get; set; } = string.Empty;                      // Konum

    [BsonElement("ownerId")]
    public string IlanSahibi { get; set; } = string.Empty;                       // İlan Sahibi

    [BsonElement("paintAndChanged")]
    public BoyaliveDegisen BoyaliDegisen { get; set; } = new();             // Boya & Değişen

    [BsonElement("createdAt")]
    public DateTime IlanTarihi { get; set; } = DateTime.UtcNow;               // İlan Tarihi
}

File: otopusula_mobile/lib/data/models/car.dart
class CarLocation {
  final String city;
  final String district;

  const CarLocation({required this.city, required this.district});

  // Backend konum alanı "Şehir, İlçe" formatında tek string döner
  factory CarLocation.fromKonum(String konum) {
    final idx = konum.indexOf(', ');
    if (idx == -1) return CarLocation(city: konum, district: '');
    return CarLocation(
      city: konum.substring(0, idx),
      district: konum.substring(idx + 2),
    );
  }

  factory CarLocation.fromJson(Map<String, dynamic> json) => CarLocation(
        city: json['city'] as String? ?? '',
        district: json['district'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {'city': city, 'district': district};

  String toKonum() => district.isEmpty ? city : '$city, $district';

  CarLocation copyWith({String? city, String? district}) =>
      CarLocation(city: city ?? this.city, district: district ?? this.district);
}

class Car {
  final String id;
  final String userId;
  final String brand;
  final String series;
  final String model;
  final int year;
  final int km;
  final String fuelType;
  final String gearType;
  final String bodyType;
  final String color;
  final int price;
  final CarLocation location;
  final String description;
  final List<String> images;
  final List<String> damageInfo;
  final bool heavilyDamaged;
  final bool tradeEligible;
  final String sellerType;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Car({
    required this.id,
    required this.userId,
    required this.brand,
    this.series = '',
    required this.model,
    required this.year,
    required this.km,
    required this.fuelType,
    required this.gearType,
    this.bodyType = '',
    this.color = '',
    required this.price,
    required this.location,
    this.description = '',
    required this.images,
    this.damageInfo = const [],
    this.heavilyDamaged = false,
    this.tradeEligible = false,
    this.sellerType = '',
    required this.createdAt,
    this.updatedAt,
  });

  // Backend: {id, marka, seri, model, yil, fiyat, kilometre, vitesTipi, yakitTipi,
  //           kasaTipi, renk, ilanSahibi, konum, resimler, boyaliDegisen,
  //           agirHasarKaydi, takasaUygun, kimden, ilanTarihi}
  factory Car.fromJson(Map<String, dynamic> json) {
    final konum = json['konum'] as String? ?? '';
    final location = CarLocation.fromKonum(konum);

    // boyaliDegisen: Orijinal olmayan panelleri hasar listesine çevir
    final boyali = json['boyaliDegisen'] as Map<String, dynamic>?;
    final damageInfo = <String>[];
    if (boyali != null) {
      boyali.forEach((panel, status) {
        if (status != null && status.toString() != 'Orijinal') {
          damageInfo.add('$panel: $status');
        }
      });
    }

    return Car(
      id: json['id'] as String,
      userId: json['ilanSahibi'] as String? ?? '',
      brand: json['marka'] as String? ?? '',
      series: json['seri'] as String? ?? '',
      model: json['model'] as String? ?? '',
      year: (json['yil'] as num?)?.toInt() ?? 0,
      km: (json['kilometre'] as num?)?.toInt() ?? 0,
      fuelType: json['yakitTipi'] as String? ?? '',
      gearType: json['vitesTipi'] as String? ?? '',
      bodyType: json['kasaTipi'] as String? ?? '',
      color: json['renk'] as String? ?? '',
      price: (json['fiyat'] as num?)?.toInt() ?? 0,
      location: location,
      description: json['aciklama'] as String? ?? '',
      images: (json['resimler'] as List<dynamic>? ?? []).cast<String>(),
      damageInfo: damageInfo,
      heavilyDamaged: json['agirHasarKaydi'] as bool? ?? false,
      tradeEligible: json['takasaUygun'] as bool? ?? false,
      sellerType: json['kimden'] as String? ?? '',
      createdAt: json['ilanTarihi'] != null
          ? DateTime.parse(json['ilanTarihi'] as String)
          : DateTime.now(),
      updatedAt: json['guncellemeTarihi'] != null
          ? DateTime.parse(json['guncellemeTarihi'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'ilanSahibi': userId,
        'marka': brand,
        'seri': series,
        'model': model,
        'yil': year,
        'kilometre': km,
        'yakitTipi': fuelType,
        'vitesTipi': gearType,
        'kasaTipi': bodyType,
        'renk': color,
        'fiyat': price,
        'konum': location.toKonum(),
        'aciklama': description,
        'resimler': images,
        'agirHasarKaydi': heavilyDamaged,
        'takasaUygun': tradeEligible,
        'kimden': sellerType,
        'ilanTarihi': createdAt.toIso8601String(),
      };

  Car copyWith({
    String? id,
    String? userId,
    String? brand,
    String? series,
    String? model,
    int? year,
    int? km,
    String? fuelType,
    String? gearType,
    String? bodyType,
    String? color,
    int? price,
    CarLocation? location,
    String? description,
    List<String>? images,
    List<String>? damageInfo,
    bool? heavilyDamaged,
    bool? tradeEligible,
    String? sellerType,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) =>
      Car(
        id: id ?? this.id,
        userId: userId ?? this.userId,
        brand: brand ?? this.brand,
        series: series ?? this.series,
        model: model ?? this.model,
        year: year ?? this.year,
        km: km ?? this.km,
        fuelType: fuelType ?? this.fuelType,
        gearType: gearType ?? this.gearType,
        bodyType: bodyType ?? this.bodyType,
        color: color ?? this.color,
        price: price ?? this.price,
        location: location ?? this.location,
        description: description ?? this.description,
        images: images ?? this.images,
        damageInfo: damageInfo ?? this.damageInfo,
        heavilyDamaged: heavilyDamaged ?? this.heavilyDamaged,
        tradeEligible: tradeEligible ?? this.tradeEligible,
        sellerType: sellerType ?? this.sellerType,
        createdAt: createdAt ?? this.createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
      );
}
